nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java

```diff
@@ -0,0 +1,341 @@
+@CapabilityDescription("Subscribes to a topic and receives messages from an MQTT broker")
+@SeeAlso({PublishMQTT.class})
+@WritesAttributes({
+    @WritesAttribute(attribute=BROKER_ATTRIBUTE_KEY, description="MQTT broker that was the message source"),
+    @WritesAttribute(attribute=TOPIC_ATTRIBUTE_KEY, description="MQTT topic on which message was received"),
+    @WritesAttribute(attribute=QOS_ATTRIBUTE_KEY, description="The quality of service for this message."),
+    @WritesAttribute(attribute=IS_DUPLICATE_ATTRIBUTE_KEY, description="Whether or not this message might be a duplicate of one which has already been received."),
+    @WritesAttribute(attribute=IS_RETAINED_ATTRIBUTE_KEY, description="Whether or not this message was from a current publisher, or was \"retained\" by the server as the last message published " +
+            "on the topic.")})
+public class ConsumeMQTT extends AbstractMQTTProcessor {
+    public final static String BROKER_ATTRIBUTE_KEY =  "mqtt.broker";
+    public final static String TOPIC_ATTRIBUTE_KEY =  "mqtt.topic";
+    public final static String QOS_ATTRIBUTE_KEY =  "mqtt.qos";
+    public final static String IS_DUPLICATE_ATTRIBUTE_KEY =  "mqtt.isDuplicate";
+    public final static String IS_RETAINED_ATTRIBUTE_KEY =  "mqtt.isRetained";
+
+    public static final PropertyDescriptor PROP_TOPIC_FILTER = new PropertyDescriptor.Builder()
+            .name("Topic Filter")
+            .description("The MQTT topic filter to designate the topics to subscribe to.")
+            .required(true)
+            .expressionLanguageSupported(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_QOS = new PropertyDescriptor.Builder()
+            .name("Quality of Service(QoS)")
+            .description("The Quality of Service(QoS) to receive the message with. Accepts values '0', '1' or '2'; '0' for 'at most once', '1' for 'at least once', '2' for 'exactly once'.")
+            .required(true)
+            .defaultValue(ALLOWABLE_VALUE_QOS_0.getValue())
+            .allowableValues(
+                    ALLOWABLE_VALUE_QOS_0,
+                    ALLOWABLE_VALUE_QOS_1,
+                    ALLOWABLE_VALUE_QOS_2)
+            .build();
+
+    public static final PropertyDescriptor PROP_MAX_QUEUE_SIZE = new PropertyDescriptor.Builder()
+            .name("Max Queue Size")
+            .description("The MQTT messages are always being sent to subscribers on a topic. If the 'Run Schedule' is significantly behind the rate at which the messages are arriving to this " +
+                    "processor then a back up can occur. This property specifies the maximum number of messages this processor will hold in memory at one time.")
+            .required(true)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .build();
+
+
+    private static int DISCONNECT_TIMEOUT = 5000;
+    private long maxQueueSize;
```

Comment:
Hate to be picky, but given that this is set in @OnSchduled but used elsewhere (other threads) it needs to be _volatile_.